File: LeetCodeProblemsGo/LeetCode2359FindClosestNodeToGivenTwoNodes/LeetCode2359FindClosestNodeToGivenTwoNodes.go
package main

import (
	"math"
)

func closestMeetingNode(edges []int, node1 int, node2 int) int {
	length := len(edges)
	visited := make([]bool, length)
	distanceFromFirstNode := make([]int, length)
	distanceFromSecondNode := make([]int, length)
	for i := 0; i < length; i++ {
		distanceFromFirstNode[i] = -1
		distanceFromSecondNode[i] = -1
	}
	distanceFromFirstNode[node1] = 0
	visited[node1] = true
	dfs(edges, visited, distanceFromFirstNode, node1, 1)
	visited = make([]bool, length)
	visited[node2] = true
	distanceFromSecondNode[node2] = 0
	dfs(edges, visited, distanceFromSecondNode, node2, 1)
	minMaxDistance := math.MaxInt32
	var destNode int
	for i := 0; i < length; i++ {
		if distanceFromFirstNode[i] == -1 {
			continue
		}
		if distanceFromSecondNode[i] == -1 {
			continue
		}
		maxDistance := distanceFromFirstNode[i]
		if distanceFromSecondNode[i] > maxDistance {
			maxDistance = distanceFromSecondNode[i]
		}
		if maxDistance < minMaxDistance {
			minMaxDistance = maxDistance
			destNode = i
		}
	}
	if minMaxDistance == math.MaxInt32 {
		return -1
	}
	return destNode
}
func dfs(edges []int, visited []bool, distancesFrom []int, node int, distance int) {
	next := edges[node]
	if next == -1 {
		return
	}
	if visited[next] {
		return
	}
	visited[next] = true
	distancesFrom[next] = distance
	dfs(edges, visited, distancesFrom, next, distance+1)
}

File: LeetCodeProblemsCMake/Problems/LeetCode2359FindClosestNodeToGivenTwoNodes.cpp
//
// Created by saeed on 5/30/25.
//

#include "LeetCode2359FindClosestNodeToGivenTwoNodes.h"

#include <iostream>
#include <limits>
#include <unordered_map>
using namespace std;

void Dfs(vector<int> &edges, vector<bool> &visited, vector<int> &distances, int node, int distance) {
    auto next = edges[node];
    if (next == -1||visited[next] )
        return;
    visited[next] = true;
    distances[next] = distance;
    Dfs(edges, visited, distances, next, distance + 1);
}

int LeetCode2359FindClosestNodeToGivenTwoNodes::closestMeetingNode(vector<int> &edges, int node1, int node2) {
    int length = static_cast<int>(edges.size());
    vector<bool> visited(length, false);
    vector<int> distanceToFirstNode(length, -1);
    vector<int> distanceToSecondNode(length, -1);
    visited[node1] = true;
    distanceToFirstNode[node1] = 0;
    Dfs(edges, visited, distanceToFirstNode, node1, 1);
    visited = vector<bool>(length, false);
    visited[node2] = true;
    distanceToSecondNode[node2] = 0;
    Dfs(edges, visited, distanceToSecondNode, node2, 1);
    int maxInt = numeric_limits<int>::max();
    int minMaxDistance = maxInt;
    int destNode{};
    for (int i = 0; i < length; i++) {
        if (distanceToFirstNode[i] < 0)
            continue;
        if (distanceToSecondNode[i] < 0)
            continue;
        auto maxDistance = std::max(distanceToFirstNode[i], distanceToSecondNode[i]);
        if (maxDistance < minMaxDistance) {
            minMaxDistance = maxDistance;
            destNode = i;
        }
    }
    if (minMaxDistance == maxInt)
        return -1;
    return destNode;
}

void LeetCode2359FindClosestNodeToGivenTwoNodes::Run() {
    vector<int>edges{4,4,4,5,1,2,2};
    int node1=1;
    int node2=1;
    std::cout<<closestMeetingNode(edges, node1, node2);
}
